lib/loader/index.js

```diff
@@ -193,10 +193,12 @@ function postInstantiate(extendedExports, instance) {
   }
 
   /** Allocates a new array in the module's memory and returns its pointer. */
-  function __newArray(id, values) {
+  function __newArray(id, valuesOrCapacity) {
+    const input = valuesOrCapacity;
```

Comment:
```suggestion
    const length = typeof valuesOrCapacity === "number" ? valuesOrCapacity : valuesOrCapacity.length;
```

Instead of the `arrayLike` logic below, perhaps a straight forward `length != valuesOrCapacity`?

Reply:
we still need to know is it array like or not for skipping array's init part: https://github.com/AssemblyScript/assemblyscript/pull/2024/files#diff-05ea19c38727a0314a536ca64b40a4fb00924bac7ffddff6a7bc3e5ac47510c4R217

Reply:
My thought was that it's array-like when `length !== valuesOrCapacity`.

Reply:
I decided use better approach and check `typeof input[Symbol.iterator] === 'function'`. This should allow use even `__newArray(ID, { length: 23 })`

Reply:
Can you elaborate on where this can be useful? I don't quite see a need for iterator and an object with a length property tbh :)

Reply:
Well it's just usual pattern. Like `Array.from({ length: 24 })`

Reply:
May I suggest to keep it simple for now? We could still add it later if a need arises.

Reply:
In this case we should make more precise type definitions. `ArrayLike<T> | number` is not proper way. It should be `Array<T> | ArrayBufferView | number`

Reply:
Basically, I think `__newArray(ID, { length: 123 })` even better and more idiomatic than `__newArray(ID, 123)`. So maybe it would be better just to remove number and leave only ArrayLike. Then it would even be a regular "fix" and not a "feat" PR

Reply:
ArrayLike should be fine I think, as it has both a `.length` and an index signature, which an object with just a `.length` does not have.

```ts
interface ArrayLike<T> {
    readonly length: number;
    readonly [n: number]: T;
}
```

Reply:
yeah, but ArrayLike support `{ length: 2 }`:
```ts
function newArray<T>(input: ArrayLike<T>) {}

newArray(new Array<number>()); // valid
newArray(new Uint32Array(1)); // valid
newArray({ length: 123 }); // valid
```
So As I suggesting keep `ArrayLike` but add special case for `{ length: n }`

Reply:
Also, right now an input like `{ length: 1, 0: 42 }` would unexpectedly drop the value.

Reply:
That `{ length: 123 }` matches `ArrayLike` is a little surprising, though. Perhaps it only does that to account for `{ length: 0 }` where there are zero values?

Reply:
Hmm, alright. I will keep only `Array<T> | ArrayBufferView | number` and remove iterator check